xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,393 @@
+	// + pending balancers are both populated), this update will always be
+	// forwarded to the pending. Thus, there is a guarantee that this will not
+	// break the balancer API of the balancer by updating after closing.
+	return balToUpdate.UpdateClientConnState(state)
+}
+
+func (gsb *gracefulSwitchBalancer) ResolverError(err error) {
+	// The update will be forwarded to the pending balancer only if there is a
+	// pending LB present, as that is the most recent LB Config prepared by the
+	// resolver, and thus is a separate concern from the current.
+	balToUpdate := gsb.latestBalancer()
+	if balToUpdate == nil {
+		return
+	}
+	// Forwarding the update outside of holding gsb.mu prevents deadlock
+	// scenarios, as the update can induce an inline callback which requires
+	// gsb.mu. However, the current balancer can be closed at any time after
+	// giving up gsb.mu. The pending balancer cannot be closed, as we are
+	// guaranteed grpc will call the balancer API synchronously, and for the
+	// pending to also be closed would require another update from grpc. The
+	// current being closed is not a problem for this ResolverError() call, as
+	// in the scenario that the current balancer can be closed (current +
+	// pending balancers are both populated), this error will always be
+	// forwarded to the pending. Thus, there is a guarantee that this will not
+	// break the balancer API of the balancer by updating after closing.
+	balToUpdate.ResolverError(err)
+}
+
+func (gsb *gracefulSwitchBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	// At any given time, the current balancer may be closed in the situation
+	// where there is a current balancer and a pending balancer. Thus, the
+	// swap() (close() + clearing of SubCons) operation is also guarded by this
+	// currentMu. Whether the current has been deleted or not will be checked
+	// from reading from the scToSubBalancer map, as that will be cleared in the
+	// atomic swap() operation. We are guaranteed that there can only be one
+	// balancer closed within this UpdateSubConnState() call, as multiple
+	// balancers closed would require another call to SwitchTo() which we are
+	// guaranteed won't be called concurrently. UpdateSubConnState() is
+	// different to UpdateClientConnState() and ResolverError(), as for those
+	// functions you don't update the current balancer in the situation where
+	// the current balancer could be closed (current + pending balancer
+	// populated).
+	gsb.currentMu.Lock()
+	defer gsb.currentMu.Unlock()
+	gsb.mu.Lock()
+	if gsb.closed {
+		gsb.mu.Unlock()
+		return
+	}
+	// This SubConn update will forward to the current balancer even if there is
+	// a pending present. This is because if this balancer does not forward the
+	// update, the picker from the current will not be updated with SubConns,
+	// leading to the possibility that the ClientConn constantly picks bad
+	// SubConns in the Graceful Switch period.
```

Comment:
It seems like "forwarding the subconn update to the appropriate child" should be enough of an explanation?

Reply:
Still, we can simplify:
```go
// Forward update to the appropriate child.  Even if there is a pending balancer, the current balancer
// should get subconn updates to maintain the proper state while the pending is still connecting.
```